### PR TITLE
2364 - 2

### PR DESCRIPTION
**Pull request by Vadser**
## Номер
2364

## Номер задания
2

## Ссылка на видео с демо
https://youtu.be/O4D7RmEDt2o

## Комментарии
сделал 2 уровня

### CODE REVIEW

**Comment by Vadser** [comment]
Please check

File: 2364/2/versus.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  end
+
+  def parse_name(opts)
+    opts.on('--name=Name', 'Name of raper') do |raper|
+      @name = raper
+    end
+  end
+end
+
+Versus.new
```
Comment (Saicheg) [inline 2364/2/versus.rb]: don't use initializer for anything except initializing object

correct version should be `Versus.new.run`

File: 2364/2/versus.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+require 'optparse'
+load 'parser.rb'
```
Comment (Saicheg) [inline 2364/2/versus.rb]: use require_relative instead of load

File: 2364/2/parser.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  def unknown_output(name)
+    puts "Рэпер #{name} не известен мне. Зато мне известны:"
+    participants[0...3].each { |raper| puts raper.name }
+  end
+
+  def print_obscene_participants(value)
+    rows = []
+    @participants = @participants[0..value - 1] if value
+    @participants.each { |participant| rows << participant.to_str }
+    puts Terminal::Table.new rows: rows
```
Comment (Saicheg) [inline 2364/2/parser.rb]: don't mix parser and presenter classes, you should follow single responsibility principals here

File: 2364/2/parser.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+class Parser
+  PATH = File.expand_path('.').freeze
+  attr_reader :participants, :methods
+  def initialize(value, name)
+    @participants = []
+    @methods = create_methods
+    create_participants(Dir[PATH + '/rap-battles/*'])
+    @participants.sort_by! { |participant| participant.participant_data[:bad_words] }
+    @participants.reverse!
+    if name
```
Comment (Saicheg) [inline 2364/2/parser.rb]: instead of using separate methods - it should be using separate classes

File: 2364/2/parser.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+load 'modules.rb'
+
+# Class parse for parsing data
+class Parser
+  PATH = File.expand_path('.').freeze
+  attr_reader :participants, :methods
+  def initialize(value, name)
+    @participants = []
+    @methods = create_methods
+    create_participants(Dir[PATH + '/rap-battles/*'])
```
Comment (Saicheg) [inline 2364/2/parser.rb]: it's definetely not a good option to hard-code it this way, better to move it to constant outside and pass as param

File: 2364/2/modules.rb [hunk 6]
```diff
@@ -0,0 +1,78 @@
+# Module for usefull methods
+module FunctionalMethods
```
Comment (Saicheg) [inline 2364/2/modules.rb]: you should avoid this type of methods and split them to service classes 

https://www.google.by/search?q=ruby+service+class&oq=ruby+service+class&aqs=chrome..69i57j0j69i60j69i65j69i60l2.2580j0j1&sourceid=chrome&ie=UTF-8

File: 2364/2/participant.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+  def count_words(value)
+    words = participant_data[:all_texts].downcase.delete('.,!?—«»').split(' ').reject { |word| word.size < 5 }
+    FunctionalMethods.output_words(words, value)
+  end
+
+  def to_str
+    bad_words = participant_data[:bad_words]
+    average_words = participant_data[:words_amount] / participant_data[:rounds]
+    average_bad_words = (bad_words.to_f / battles).round(2)
```
Comment (Saicheg) [inline 2364/2/participant.rb]: what will happen if battles will equal 0?

File: 2364/2/parser.rb (hunk 4 above)
Comment (Vadser) [reply 2364/2/parser.rb]: I do not fully understand what this means

**Comment by Vadser** [comment]
Recheck, please

**Comment by Vadser** [comment]
Recheck, please

File: 2364/2/name_check.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
+# Class for comparing names
+class NameCheck
```
Comment (Saicheg) [inline 2364/2/name_check.rb]: class names should always be verbs, if they are doing something

`NameChacker.new.run` in this case

File: 2364/2/parser.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
+require 'russian_obscenity'
+require 'terminal-table'
+require_relative 'participant.rb'
```
Comment (Saicheg) [inline 2364/2/parser.rb]: you don't need `.rb` part

File: 2364/2/spec/name_check_test_spec.rb [hunk 10]
```diff
@@ -0,0 +1,17 @@
+require_relative '../parser.rb'
+require_relative '../participant.rb'
+require_relative '../count_words.rb'
+require_relative '../name_check.rb'
+require_relative '../create_data.rb'
+
+describe NameCheck do
+  it 'check if class of name_check NameCheck' do
+    name_check = NameCheck.new('', '')
+    expect(NameCheck).to eq name_check.class
```
Comment (Saicheg) [inline 2364/2/spec/name_check_test_spec.rb]: this is always be true

File: 2364/2/spec/name_check_test_spec.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  end
+
+  it 'check two names' do
+    check = NameCheck.new('Oxxxymiron', "Oxxxymiron'a").name_check
+    expect(true).to eq check
```
Comment (Saicheg) [inline 2364/2/spec/name_check_test_spec.rb]:
```
expect(check) to eq(true)

File: 2364/2/parser.rb [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+require_relative 'top_words.rb'
+
+# Class parse for parsing data
+class Parser
+  PATH = File.expand_path('.').freeze
+  attr_reader :participants, :methods
+  def initialize
+    @participants = []
+    @methods = create_methods
+    files_path = Dir[PATH + '/rap-battles/*']
```
Comment (Saicheg) [inline 2364/2/parser.rb]: this should be passed as param from main class

File: 2364/2/parser.rb [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+require_relative 'participant.rb'
+require_relative 'create_data.rb'
+require_relative 'name_check.rb'
+require_relative 'obscene_rapers.rb'
+require_relative 'top_words.rb'
+
+# Class parse for parsing data
+class Parser
+  PATH = File.expand_path('.').freeze
+  attr_reader :participants, :methods
```
Comment (Saicheg) [inline 2364/2/parser.rb]: i am not sure, but i think this needs to be separated with additional spaces above

File: 2364/2/parser.rb [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    if name
+      TopWords.new(value, name, participants).print_top_words
+    else
+      ObsceneRapers.new(value, participants).print_obscene_participants
+    end
+  end
+
+  private
+
+  def create_methods
```
Comment (Saicheg) [inline 2364/2/parser.rb]: this is black magic, get rid of that proc logic or split it to separate classes and cover with tests

File: 2364/2/parser.rb [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    hash_methods = Hash.new(proc { |obj, data| obj.update_data(data) })
+    hash_methods[nil] = proc { |*args| @participants << Participant.new(args[0], args[1]) }
+    hash_methods
+  end
+
+  def create_participants(file_names)
+    file_names.each do |file_name|
+      data = CreateData.new(file_name).take_data
+      name = data[:name]
+      check_participant(data, name)
```
Comment (Saicheg) [inline 2364/2/parser.rb]: check_participant for what?

File: 2364/2/parser.rb [hunk 16]
```diff
@@ -0,0 +1,48 @@
+require 'russian_obscenity'
+require 'terminal-table'
+require_relative 'participant.rb'
+require_relative 'create_data.rb'
+require_relative 'name_check.rb'
+require_relative 'obscene_rapers.rb'
+require_relative 'top_words.rb'
+
+# Class parse for parsing data
+class Parser
```
Comment (Saicheg) [inline 2364/2/parser.rb]: Parser of what?

**Comment by Vadser** [comment]
@Saicheg, check, please

**Pull request merged**